### PR TITLE
Training and Finetuning Reranker Models with Sentence Transformers v4

### PR DESCRIPTION
**Pull request by tomaarsen**
This blogpost is designed to release alongside the v4.0 Sentence Transformers update, ideally on or around March 27th.

It's a blogpost-ification of the "Training Overview" documentation that I'll be publishing when v4.0 is out. That's why the format is a bit odd for a blogpost. I did the same for v3.0 with the rather successful https://huggingface.co/blog/train-sentence-transformers.

It looks like I actually made a pretty cool model when trying to explain how to finetuning reranker models:
reranker_gooaq_model_size_ndcg

## Preparing the Article

You're not quite done yet, though. Please make sure to follow this process (as documented [here](https://github.com/huggingface/blog/tree/main?tab=readme-ov-file#how-to-write-an-article-)):

- [x] Add an entry to [_blog.yml](https://github.com/huggingface/blog/blob/main/_blog.yml).
- [x] Add a thumbnail. There are no requirements here, but there is a [template](https://github.com/huggingface/blog?tab=readme-ov-file#how-to-get-a-nice-responsive-thumbnail) if it's helpful.
- [x] Check you use a short title and blog path.
- [x] Upload any additional assets (such as images) to the Documentation Images [repo](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog). This is to reduce bloat in the GitHub base repo when cloning and pulling. Try to have small images to avoid a slow or expensive user experience.
- [x] Add metadata (such as authors) to your `md` file. You can also specify `guest` or `org` for the authors.
- [x] Ensure the publication date is correct.
- [x] Preview the content. A quick way is to paste the markdown content in https://huggingface.co/new-blog. **Do not click publish**, this is just a way to do an early check.

Here is an example of a complete PR: https://github.com/huggingface/blog/pull/2382

## Getting a Review

Please make sure to get a review from someone on your team or a co-author.
Once this is done and once all the steps above are completed, you should be able to merge.
There is no need for additional reviews if you and your co-authors are happy and meet all of the above.

Feel free to add @pcuenca as a reviewer if you want a final check. Keep in mind he'll be biased toward light reviews
(e.g., check for proper metadata) rather than content reviews unless explicitly asked. 

cc @LysandreJik to keep you in the loop

- Tom Aarsen

### CODE REVIEW

File: train-reranker.md [hunk 1]
```diff
@@ -0,0 +1,743 @@
...
+- [Training Tips](#training-tips)
+<!-- - [Deprecation](#deprecation) -->
+- [Evaluation](#evaluation)
+- [Additional Resources](#additional-resources)
+  * [Training Examples](#training-examples)
+  * [Documentation](#documentation)
+
+## What are Reranker models?
+
+Reranker models, often implemented using Cross Encoder architectures, are designed to evaluate the relevance between pairs of texts (e.g., a query and a document, or two sentences). Unlike Sentence Transformer (a.k.a. bi-encoders, embedding models), which independently embed each text into vectors and compute similarity via a distance metric, Cross Encoder process the paired texts together through a shared neural network, resulting in one output score. By letting the two texts attend to each other, Cross Encoder models can reach stronger performance than embedding models.
```
Comment (davanstrien) [inline train-reranker.md:48]:
```suggestion
Reranker models, often implemented using Cross Encoder architectures, are designed to evaluate the relevance between pairs of texts (e.g., a query and a document, or two sentences). Unlike Sentence Transformer (a.k.a. bi-encoders, embedding models), which independently embed each text into vectors and compute similarity via a distance metric, Cross Encoder process the paired texts together through a shared neural network, resulting in one output score. By letting the two texts attend to each other, Cross Encoder models can perform better than embedding models.
```

File: train-reranker.md [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+
+However, this strength comes with a trade-off: Cross Encoder models are slower as they process every possible pair of texts (e.g., 10 queries with 500 candidate documents requires 5,000 computations instead of 510 for embedding models). This makes them less efficient for large-scale initial retrieval but ideal for reranking: refining the top-k results first identified by faster Sentence Transformer models. This 2-stage "retrieve and rerank" approach is commonly used in the strongest search systems.
```
Comment (davanstrien) [inline train-reranker.md:50]:
```suggestion
However, this strength comes with a trade-off: Cross Encoder models are slower as they process every possible pair of texts (e.g., 10 queries with 500 candidate documents requires 5,000 computations instead of 510 for embedding models). This makes them less efficient for large-scale initial retrieval but ideal for reranking: refining the top-k results first identified by faster Sentence Transformer models. The strongest search systems commonly use this 2-stage "retrieve and rerank" approach.
```

File: train-reranker.md [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+```
+
+Throughout this blogpost, I'll use "reranker model" and "Cross Encoder model" interchangeably.
+
+## Why Finetune?
+
+Reranker models are often tasked with a challenging problem: 
+> Which of these $k$ highly-related documents answers the query the best?
+
+General-purpose reranker models are trained to perform adequately on this exact question in a wide range of domains and topics, preventing it from reaching its maximum potential on your specific domain. Through finetuning, the model can learn to focus exclusively on the domain and/or language that matters for you.
```
Comment (davanstrien) [inline train-reranker.md:79]:
```suggestion
General-purpose reranker models are trained to perform adequately on this exact question in a wide range of domains and topics, preventing it from reaching its maximum potential in your specific domain. Through finetuning, the model can learn to focus exclusively on the domain and/or language that matters to you.
```

**Review by davanstrien — APPROVED** [review]
really nice post, added some suggestions for grammar/language but feel free to ignore!

File: train-reranker.md (hunk 1 above)
Comment (tomaarsen) [reply train-reranker.md:48]:
```suggestion
Reranker models, often implemented using Cross Encoder architectures, are designed to evaluate the relevance between pairs of texts (e.g., a query and a document, or two sentences). Unlike Sentence Transformer (a.k.a. bi-encoders, embedding models), which independently embed each text into vectors and compute similarity via a distance metric, Cross Encoder process the paired texts together through a shared neural network, resulting in one output score. By letting the two texts attend to each other, Cross Encoder models can outperform embedding models.
```

I think this might be even stronger?

File: train-reranker.md [hunk 4]
```diff
@@ -0,0 +1,819 @@
...
+-->
+
+![Embedding vs Reranker Models](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/train-reranker/embedding_vs_reranker_model.png)
+
+Throughout this blogpost, I'll use "reranker model" and "Cross Encoder model" interchangeably.
+
+## Why Finetune?
+
+Reranker models are often tasked with a challenging problem: 
+> Which of these $k$ highly-related documents answers the query the best?
```
Comment (pcuenca) [inline train-reranker.md:84]:
```suggestion
> Which of these \\(k\\) highly-related documents answers the query the best?
```

Assuming you intended a LaTeX-style rendering.

File: train-reranker.md [hunk 5]
```diff
@@ -0,0 +1,819 @@
...
+
+General-purpose reranker models are trained to perform adequately on this exact question in a wide range of domains and topics, preventing it from reaching its maximum potential in your specific domain. Through finetuning, the model can learn to focus exclusively on the domain and/or language that matters to you.
```
Comment (pcuenca) [inline train-reranker.md:86]:
```suggestion
General-purpose reranker models are trained to perform adequately on this exact question in a wide range of domains and topics, preventing them from reaching their maximum potential in your specific domain. Through finetuning, the model can learn to focus exclusively on the domain and/or language that matters to you.
```

File: train-reranker.md [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+
+Additionally, if your dataset has extraneous columns (e.g. sample_id, metadata, source, type), you should remove these with [`Dataset.remove_columns`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.remove_columns) as they will be used as inputs otherwise. You can also use [`Dataset.select_columns`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.select_columns) to keep only the desired columns.
+
+### Hard Negatives Mining
+
+The success of training reranker models often depends on the quality of the *negatives*, i.e. the passages for which the query-negative score should be low. Negatives can be divided into two types:
+
+* **Soft negatives**: passages that are completely unrelated. Also called easy negatives.
+* **Hard negatives**: passages that seem like they might be relevant for the query, but are not.
```
Comment (pcuenca) [inline train-reranker.md:183]: Interesting, I thought it was the other way around ("hard" negative sounds _more_ negative).

File: train-reranker.md [hunk 7]
```diff
@@ -0,0 +1,819 @@
...
+
+The strongest CrossEncoder models are generally trained to recognize hard negatives, and so it's valuable to be able to "mine" hard negatives. Sentence Transformers supports a strong [`mine_hard_negatives`](https://sbert.net/docs/package_reference/util.html#sentence_transformers.util.mine_hard_negatives) function that can assist, given a dataset of query-answer pairs:
+
+```python
+from datasets import load_dataset
+from sentence_transformers import SentenceTransformer
+from sentence_transformers.util import mine_hard_negatives
+
+# Load the GooAQ dataset: https://huggingface.co/datasets/sentence-transformers/gooaq
+train_dataset = load_dataset("sentence-transformers/gooaq", split=f"train").select(range(100_000))
```
Comment (pcuenca) [inline train-reranker.md:198]: Why not `load_dataset("sentence-transformers/gooaq", split="train[:100000")`

File: train-reranker.md [hunk 8]
```diff
@@ -0,0 +1,819 @@
...
+NDCG@10:  59.12 -> 71.35
+"""
+# {'gooaq-dev_map': 0.6728370126462222, 'gooaq-dev_mrr@10': 0.6665190476190477, 'gooaq-dev_ndcg@10': 0.7135068904582963, 'gooaq-dev_base_map': 0.5327714512001362, 'gooaq-dev_base_mrr@10': 0.5239674603174603, 'gooaq-dev_base_ndcg@10': 0.5912299141913905}
+```
+
+## Trainer
+
+The [`CrossEncoderTrainer`](https://sbert.net/docs/package_reference/cross_encoder/trainer.html#sentence_transformers.trainer.CrossEncoderTrainer) is where all previous components come together. We only have to specify the trainer with the model, training arguments (optional), training dataset, evaluation dataset (optional), loss function, evaluator (optional) and we can start training. Let’s have a look at a script where all of these components come together:
+
+```python
```
Comment (pcuenca) [inline train-reranker.md:466]: I realize this script is important, but it's a bit long, I wonder if it should still be inside a `details` section?

File: train-reranker.md [hunk 9]
```diff
@@ -0,0 +1,819 @@
...
+            f"`huggingface-cli login`, followed by loading the model using `model = CrossEncoder({final_output_dir!r})` "
+            f"and saving it using `model.push_to_hub('{run_name}')`."
+        )
+
+
+if __name__ == "__main__":
+    main()
+```
+
+In this example I'm finetuning from [`answerdotai/ModernBERT-base`](https://huggingface.co/answerdotai/ModernBERT-base), a base model that is not yet a Cross Encoder model. This generally requires more training data than finetuning an existing reranker model like [`Alibaba-NLP/gte-multilingual-reranker-base`](https://huggingface.co/Alibaba-NLP/gte-multilingual-reranker-base). I'm using 99k query-answer pairs from the [GooAQ dataset](https://huggingface.co/datasets/sentence-transformers/gooaq), after which I mine hard negatives using the [sentence-transformers/static-retrieval-mrl-en-v1](https://huggingface.co/sentence-transformers/static-retrieval-mrl-en-v1) embedding model. This results in 578k labeled pairs, i.e. 99k positive pairs (i.e. label=1) and 479k negative pairs (i.e. label=0).
```
Comment (pcuenca) [inline train-reranker.md:642]:
```suggestion
In this example I'm finetuning from [`answerdotai/ModernBERT-base`](https://huggingface.co/answerdotai/ModernBERT-base), a base model that is not yet a Cross Encoder model. This generally requires more training data than finetuning an existing reranker model like [`Alibaba-NLP/gte-multilingual-reranker-base`](https://huggingface.co/Alibaba-NLP/gte-multilingual-reranker-base). I'm using 99k query-answer pairs from the [GooAQ dataset](https://huggingface.co/datasets/sentence-transformers/gooaq), after which I mine hard negatives using the [sentence-transformers/static-retrieval-mrl-en-v1](https://huggingface.co/sentence-transformers/static-retrieval-mrl-en-v1) embedding model. This results in 578k labeled pairs: 99k positive pairs (i.e. label=1) and 479k negative pairs (i.e. label=0).
```

File: train-reranker.md [hunk 10]
```diff
@@ -0,0 +1,819 @@
...
+
+I use the `BinaryCrossEntropyLoss`, which is well suited for these labeled pairs. I also set up 2 forms of evaluation: `CrossEncoderNanoBEIREvaluator` which evaluates against the NanoBEIR benchmark and `CrossEncoderRerankingEvaluator` which evaluates the performance of reranking the top 30 results from the aforementioned static embedding model. Afterwards, I defined a fairly standard set of hyperparameters, including learning rates, warmup ratios, bf16, loading the best model at the end, and some debugging parameters. Lastly, I run the trainer, perform post-training evaluation, and save the model both locally and on the Hugging Face Hub.
```
Comment (pcuenca) [inline train-reranker.md:644]:
```suggestion
I use the `BinaryCrossEntropyLoss`, which is well suited for these labeled pairs. I also set up 2 forms of evaluation: `CrossEncoderNanoBEIREvaluator` which evaluates against the NanoBEIR benchmark and `CrossEncoderRerankingEvaluator` which evaluates the performance of reranking the top 30 results from the aforementioned static embedding model. Afterwards, I define a fairly standard set of hyperparameters, including learning rates, warmup ratios, bf16, loading the best model at the end, and some debugging parameters. Lastly, I run the trainer, perform post-training evaluation, and save the model both locally and on the Hugging Face Hub.
```

File: train-reranker.md (hunk 4 above)
Comment (tomaarsen) [reply train-reranker.md:84]: I did! I wasn't aware of this notation, thanks!

File: train-reranker.md (hunk 6 above)
Comment (tomaarsen) [reply train-reranker.md:183]: It's a bit of a tricky definition, hence the examples. "hard negative" means "hard for a model to see as a negative". 

One of the original formal definitions actually says that a "hard negative" is any negative that is deemed more similar to an anchor (e.g. query) than the positive, but there's a very blurry line between "hard negative" and "positive".

File: train-reranker.md [hunk 11]
```diff
@@ -0,0 +1,819 @@
...
+
+* **Soft negatives**: passages that are completely unrelated. Also called easy negatives.
+* **Hard negatives**: passages that seem like they might be relevant for the query, but are not.
+
+A concise example is:
+* **Query**: Where was Apple founded?
+* **Soft Negative**: The Cache River Bridge is a Parker pony truss that spans the Cache River between Walnut Ridge and Paragould, Arkansas.
+* **Hard Negative**: The Fuji apple is an apple cultivar developed in the late 1930s, and brought to market in 1962.
+
+The strongest CrossEncoder models are generally trained to recognize hard negatives, and so it's valuable to be able to "mine" hard negatives. Sentence Transformers supports a strong [`mine_hard_negatives`](https://sbert.net/docs/package_reference/util.html#sentence_transformers.util.mine_hard_negatives) function that can assist, given a dataset of query-answer pairs:
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply train-reranker.md:190]:
```suggestion
The strongest CrossEncoder models are generally trained to recognize hard negatives, and so it's valuable to be able to "mine" hard negatives to train the with. Sentence Transformers supports a strong [`mine_hard_negatives`](https://sbert.net/docs/package_reference/util.html#sentence_transformers.util.mine_hard_negatives) function that can assist, given a dataset of query-answer pairs:
```
Your proposal has a different meaning than my intention, so that tells me that my original text wasn't clear enough.

File: train-reranker.md (hunk 7 above)
Comment (tomaarsen) [reply train-reranker.md:198]:
```suggestion
train_dataset = load_dataset("sentence-transformers/gooaq", split="train").select(range(100_000))
```
Because I people to be familiarized with `select(range(...))`, as that is useful outside of just `load_dataset`, e.g. also when you're using `Dataset.from_dict`, `Dataset.from_list`, etc. I think I actually used `f"train[:{max_length}]"` in my own script, that's why the `f` was still there.

Also, I'm pretty sure that `train[:100000]` still downloads the full dataset and only afterwards does the slicing.

File: train-reranker.md (hunk 8 above)
Comment (tomaarsen) [reply train-reranker.md:466]: I've also been thinking about this, I'll do some brainstorming what I prefer.

File: train-reranker.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply train-reranker.md]: ah I see what you meant, sorry!

[the remark this answers is not in the document]
Comment (tomaarsen) [reply train-reranker.md]: No need to apologize! It's very useful to show me that the original intention was not clear 🤗

File: train-reranker.md (hunk 8 above)
Comment (tomaarsen) [reply train-reranker.md:466]: My preference is to keep the script to try and help familiarize potential users with the code structure. I care more about trying to get people to make their own models than to focus on the 2 models that I trained for the blogpost, and I think having the code (hopefully) helps with that.
